libraries-3/src/main/java/com/baeldung/jnats/NatsClient.java

```diff
@@ -1,123 +1,144 @@
 package com.baeldung.jnats;
 
 import java.io.IOException;
+import java.time.Duration;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
 
+import io.nats.client.*;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import io.nats.client.AsyncSubscription;
-import io.nats.client.Connection;
-import io.nats.client.Message;
-import io.nats.client.Nats;
-import io.nats.client.Options;
-import io.nats.client.Subscription;
-import io.nats.client.SyncSubscription;
-
 public final class NatsClient {
 
     private final String serverURI;
 
-    private final Connection natsConnection;
-
-    private final Map<String, Subscription> subscriptions = new HashMap<>();
+    private final Map<Subscription, Dispatcher> dispatcherBySubscription = new HashMap<>();
 
     private final static Logger log = LoggerFactory.getLogger(NatsClient.class);
 
+    private Connection natsConnection;
+
     NatsClient() {
-        this.serverURI = "jnats://localhost:4222";
-        natsConnection = initConnection(serverURI);
+        this(null);
     }
 
     public NatsClient(String serverURI) {
-        if ((serverURI != null) && (!serverURI.isEmpty())) {
+        if (serverURI == null || serverURI.isEmpty()) {
+            this.serverURI = "nats://localhost:4222";
+        }
+        else {
             this.serverURI = serverURI;
-        } else {
-            this.serverURI = "jnats://localhost:4222";
         }
+        initConnection();
+    }
```

Comment:
I'm not a fan of having behaviour/logic inside constructors, can we refactor this to use a static factory method and keep the constructor responsibility just for assigning the fields

Reply:
Can you give me a better explanation? Maybe the question is would you have made the same comment on the original code, because it's pretty much the same.

Reply:
Fixed.

Reply:
Sorry I missed your question here. I mean, the single responsibility of a constructor is usually simply to assign fields.

Putting if statements, or other init logic inside a constructor is generally considered bad practice and comes as a result of a lack of proper dependency injection. I'll leave a suggestion showing what I would usually expect to see here.